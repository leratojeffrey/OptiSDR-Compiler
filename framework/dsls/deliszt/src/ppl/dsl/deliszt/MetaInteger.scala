package ppl.dsl.deliszt

/**
 * author: Michael Wu (mikemwu@stanford.edu)
 * last modified: 04/28/2011
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

/*
 * MetaInteger.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

abstract class IntM {
  type Suc <: IntM
  type Add[N <: IntM] <: IntM
}

object EnsureSize {
  implicit def EnsureSizeZero[T <: IntM] =
  new EnsureSize[Zero,Succ[T]] {
    def get(n : Int) : Int = n
  }

  implicit def EnsureSizeSucc[T <: IntM, TT <: IntM](implicit fn : EnsureSize[TT,T]) =
  new EnsureSize[Succ[TT],Succ[T]] {
    def get(n : Int) : Int = fn.get(n+1)
  }
}

//meta-function(?)(what is the term for this thing?)
//if IDX is less than SIZE, then an implicit object of type EnsureSize[IDX,SIZE] will
//exist with a idx function that returns the number IDX as an integer
abstract class EnsureSize[TT <: IntM,T <: IntM] {
  def get(n : Int) : Int
  def idx = get(0)
}

object MVal {
  implicit val MValZero = new MVal[Zero] {
    val value = 0
  }

  implicit def MValSucc[T <: IntM](implicit f: MVal[T]) = new MVal[Succ[T]] {
    val value = f.value + 1
  }
}

abstract class MVal[T] {
  val value: Int
}

class Zero extends IntM {
  type Suc = Succ[Zero]
  type Add[N <: IntM] = N
}

class Succ[N <: IntM] extends IntM {
  type Suc = Succ[Succ[N]]
  type Add[M <: IntM] = (N#Add[M])#Suc
}

object MetaInteger {
  type +[N1 <: IntM, N2 <: IntM] = N1#Add[N2]

  //Scala doesn't provide meta-integers like C++'s typename<int N>
  //so we have to define some integers here
  //this is a simplification of:
  //http://michid.wordpress.com/code/meta-programming-with-scala-part-i-addition/
  //somelike like Church encoding but the numbers are themselves objects and not
  //functions for simplicity since we don't need to define addition on them
  //as far I can tell ... if necessary I suppose we can define multiplcation
  //on these if we needed numbers higher than 9
  type _0 = Zero
  type _1 = Succ[_0]
  type _2 = Succ[_1]
  type _3 = Succ[_2]
  type _4 = Succ[_3]
  type _5 = Succ[_4]
  type _6 = Succ[_5]
  type _7 = Succ[_6]
  type _8 = Succ[_7]
  type _9 = Succ[_8]
  type _10 = Succ[_9]
  type _11 = Succ[_10]
  type _12 = Succ[_11]
  type _13 = Succ[_12]
  type _14 = Succ[_13]
  type _15 = Succ[_14]
  type _16 = Succ[_15]
  type _17 = Succ[_16]
  type _18 = Succ[_17]
  type _19 = Succ[_18]
  type _20 = Succ[_19]
  type _21 = Succ[_20]
  type _22 = Succ[_21]
  
  //for convience we provide objects representing each number as well
  //so you can avoid passing parameter arguments
  val _0 : _0 = new _0
  val _1 : _1 = new _1
  val _2 : _2 = new _2
  val _3 : _3 = new _3
  val _4 : _4 = new _4
  val _5 : _5 = new _5
  val _6 : _6 = new _6
  val _7 : _7 = new _7
  val _8 : _8 = new _8
  val _9 : _9 = new _9
  val _10 : _10 = new _10
  val _11 : _11 = new _11
  val _12 : _12 = new _12
  val _13 : _13 = new _13
  val _14 : _14 = new _14
  val _15 : _15 = new _15
  val _16 : _16 = new _16
  val _17 : _17 = new _17
  val _18 : _18 = new _18
  val _19 : _19 = new _19
  val _20 : _20 = new _20
  val _21 : _21 = new _21
  val _22 : _22 = new _22

  def MIntDepth[T <: IntM]()(implicit n: MVal[T]) = n.value
  def MIntDepth[T <: IntM](m : T)(implicit n: MVal[T]) = n.value
}
