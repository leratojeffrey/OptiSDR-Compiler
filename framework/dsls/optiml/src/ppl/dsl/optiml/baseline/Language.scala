package ppl.dsl.optiml.baseline

object LanguageDefs {
  val INITIAL_SEED = 100
  var randRef = new scala.util.Random(INITIAL_SEED)
  var _time = 0L
  
  def random(): Double = randRef.nextDouble
    
  def abs(d: Double) = implicitly[Numeric[Double]].abs(d) //java.lang.Math.abs(d)
  
  def sum[B:Numeric:Manifest](st: Int, en: Int)(f: Int => B) = {
    var acc = f(st)
    for (i <- st+1 until en) {
      acc = implicitly[Numeric[B]].plus(acc,f(i))
    }
    acc
  }
  
  def tic() {
    _time = System.currentTimeMillis()
  }
  
  def toc() {
    val now = System.currentTimeMillis()
    println("elapsed: " + (now-_time)/1000.0)
  }
}