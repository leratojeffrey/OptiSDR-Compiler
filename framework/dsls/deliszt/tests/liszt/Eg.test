import Liszt.Language._
import Liszt.MetaInteger._
@lisztcode
object Eg {
    val interior = BoundarySet[Face]("default-interior")
    val face_field = FieldWithConst[Face,Int](0)
    val cell_field = FieldWithConst[Cell,Int](0)
    def main() {
        
       var global = 0
        for(f <- interior) {
            global += ID(f)
            face_field(f) = ID(f)
        }
        for(c <- cells(mesh))
           for(f <- faces(c))
            cell_field(c) += face_field(f)
        
        for(c <- cells(mesh))
          Print(cell_field(c))
    }
}