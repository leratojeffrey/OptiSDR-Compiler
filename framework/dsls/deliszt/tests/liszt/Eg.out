2:c1-0 liszt: 753
2:c2-0 liszt: 1012
2:c3-0 liszt: 1024
2:c4-0 liszt: 1036
2:c5-0 liszt: 784
2:c6-0 liszt: 1046
2:c7-0 liszt: 1322
2:c8-0 liszt: 1337
2:c9-0 liszt: 1352
2:c10-0 liszt: 1088
2:c11-0 liszt: 1102
2:c12-0 liszt: 1392
2:c13-0 liszt: 1407
2:c14-0 liszt: 1422
2:c15-0 liszt: 1144
2:c16-0 liszt: 1158
2:c17-0 liszt: 1462
2:c18-0 liszt: 1477
2:c19-0 liszt: 1492
2:c20-0 liszt: 1200
2:c21-0 liszt: 1214
2:c22-0 liszt: 1532
2:c23-0 liszt: 1547
2:c24-0 liszt: 1562
2:c25-0 liszt: 1256
2:c26-0 liszt: 1270
2:c27-0 liszt: 1602
2:c28-0 liszt: 1617
2:c29-0 liszt: 1632
2:c30-0 liszt: 1312
2:c31-0 liszt: 1326
2:c32-0 liszt: 1672
2:c33-0 liszt: 1687
2:c34-0 liszt: 1702
2:c35-0 liszt: 1368
2:c36-0 liszt: 1382
2:c37-0 liszt: 1742
2:c38-0 liszt: 1757
2:c39-0 liszt: 1772
2:c40-0 liszt: 1424
2:c41-0 liszt: 1438
2:c42-0 liszt: 1812
2:c43-0 liszt: 1827
2:c44-0 liszt: 1842
2:c45-0 liszt: 1480
2:c46-0 liszt: 1116
2:c47-0 liszt: 1499
2:c48-0 liszt: 1508
2:c49-0 liszt: 1517
2:c50-0 liszt: 1140
2:c51-0 liszt: 1410
2:c52-0 liszt: 1807
2:c53-0 liszt: 1822
2:c54-0 liszt: 1837
2:c55-0 liszt: 1452
2:c56-0 liszt: 1852
2:c57-0 liszt: 2266
2:c58-0 liszt: 2284
2:c59-0 liszt: 2302
2:c60-0 liszt: 1905
2:c61-0 liszt: 1922
2:c62-0 liszt: 2350
2:c63-0 liszt: 2368
2:c64-0 liszt: 2386
2:c65-0 liszt: 1975
2:c66-0 liszt: 1992
2:c67-0 liszt: 2434
2:c68-0 liszt: 2452
2:c69-0 liszt: 2470
2:c70-0 liszt: 2045
2:c71-0 liszt: 2062
2:c72-0 liszt: 2518
2:c73-0 liszt: 2536
2:c74-0 liszt: 2554
2:c75-0 liszt: 2115
2:c76-0 liszt: 2132
2:c77-0 liszt: 2602
2:c78-0 liszt: 2620
2:c79-0 liszt: 2638
2:c80-0 liszt: 2185
2:c81-0 liszt: 2202
2:c82-0 liszt: 2686
2:c83-0 liszt: 2704
2:c84-0 liszt: 2722
2:c85-0 liszt: 2255
2:c86-0 liszt: 2272
2:c87-0 liszt: 2770
2:c88-0 liszt: 2788
2:c89-0 liszt: 2806
2:c90-0 liszt: 2325
2:c91-0 liszt: 2342
2:c92-0 liszt: 2854
2:c93-0 liszt: 2872
2:c94-0 liszt: 2890
2:c95-0 liszt: 2395
2:c96-0 liszt: 1898
2:c97-0 liszt: 2418
2:c98-0 liszt: 2429
2:c99-0 liszt: 2440
2:c100-0 liszt: 1929
2:c101-0 liszt: 1950
2:c102-0 liszt: 2482
2:c103-0 liszt: 2497
2:c104-0 liszt: 2512
2:c105-0 liszt: 1992
2:c106-0 liszt: 2527
2:c107-0 liszt: 3076
2:c108-0 liszt: 3094
2:c109-0 liszt: 3112
2:c110-0 liszt: 2580
2:c111-0 liszt: 2597
2:c112-0 liszt: 3160
2:c113-0 liszt: 3178
2:c114-0 liszt: 3196
2:c115-0 liszt: 2650
2:c116-0 liszt: 2667
2:c117-0 liszt: 3244
2:c118-0 liszt: 3262
2:c119-0 liszt: 3280
2:c120-0 liszt: 2720
2:c121-0 liszt: 2737
2:c122-0 liszt: 3328
2:c123-0 liszt: 3346
2:c124-0 liszt: 3364
2:c125-0 liszt: 2790
2:c126-0 liszt: 2807
2:c127-0 liszt: 3412
2:c128-0 liszt: 3430
2:c129-0 liszt: 3448
2:c130-0 liszt: 2860
2:c131-0 liszt: 2877
2:c132-0 liszt: 3496
2:c133-0 liszt: 3514
2:c134-0 liszt: 3532
2:c135-0 liszt: 2930
2:c136-0 liszt: 2947
2:c137-0 liszt: 3580
2:c138-0 liszt: 3598
2:c139-0 liszt: 3616
2:c140-0 liszt: 3000
2:c141-0 liszt: 3017
2:c142-0 liszt: 3664
2:c143-0 liszt: 3682
2:c144-0 liszt: 3700
2:c145-0 liszt: 3070
2:c146-0 liszt: 2438
2:c147-0 liszt: 3093
2:c148-0 liszt: 3104
2:c149-0 liszt: 3115
2:c150-0 liszt: 2469
2:c151-0 liszt: 2490
2:c152-0 liszt: 3157
2:c153-0 liszt: 3172
2:c154-0 liszt: 3187
2:c155-0 liszt: 2532
2:c156-0 liszt: 3202
2:c157-0 liszt: 3886
2:c158-0 liszt: 3904
2:c159-0 liszt: 3922
2:c160-0 liszt: 3255
2:c161-0 liszt: 3272
2:c162-0 liszt: 3970
2:c163-0 liszt: 3988
2:c164-0 liszt: 4006
2:c165-0 liszt: 3325
2:c166-0 liszt: 3342
2:c167-0 liszt: 4054
2:c168-0 liszt: 4072
2:c169-0 liszt: 4090
2:c170-0 liszt: 3395
2:c171-0 liszt: 3412
2:c172-0 liszt: 4138
2:c173-0 liszt: 4156
2:c174-0 liszt: 4174
2:c175-0 liszt: 3465
2:c176-0 liszt: 3482
2:c177-0 liszt: 4222
2:c178-0 liszt: 4240
2:c179-0 liszt: 4258
2:c180-0 liszt: 3535
2:c181-0 liszt: 3552
2:c182-0 liszt: 4306
2:c183-0 liszt: 4324
2:c184-0 liszt: 4342
2:c185-0 liszt: 3605
2:c186-0 liszt: 3622
2:c187-0 liszt: 4390
2:c188-0 liszt: 4408
2:c189-0 liszt: 4426
2:c190-0 liszt: 3675
2:c191-0 liszt: 3692
2:c192-0 liszt: 4474
2:c193-0 liszt: 4492
2:c194-0 liszt: 4510
2:c195-0 liszt: 3745
2:c196-0 liszt: 2978
2:c197-0 liszt: 3768
2:c198-0 liszt: 3779
2:c199-0 liszt: 3790
2:c200-0 liszt: 3009
2:c201-0 liszt: 2238
2:c202-0 liszt: 3035
2:c203-0 liszt: 3044
2:c204-0 liszt: 3053
2:c205-0 liszt: 2262
2:c206-0 liszt: 3061
2:c207-0 liszt: 3869
2:c208-0 liszt: 3880
2:c209-0 liszt: 3891
2:c210-0 liszt: 3092
2:c211-0 liszt: 3102
2:c212-0 liszt: 3919
2:c213-0 liszt: 3930
2:c214-0 liszt: 3941
2:c215-0 liszt: 3133
2:c216-0 liszt: 3143
2:c217-0 liszt: 3969
2:c218-0 liszt: 3980
2:c219-0 liszt: 3991
2:c220-0 liszt: 3174
2:c221-0 liszt: 3184
2:c222-0 liszt: 4019
2:c223-0 liszt: 4030
2:c224-0 liszt: 4041
2:c225-0 liszt: 3215
2:c226-0 liszt: 3225
2:c227-0 liszt: 4069
2:c228-0 liszt: 4080
2:c229-0 liszt: 4091
2:c230-0 liszt: 3256
2:c231-0 liszt: 3266
2:c232-0 liszt: 4119
2:c233-0 liszt: 4130
2:c234-0 liszt: 4141
2:c235-0 liszt: 3297
2:c236-0 liszt: 3307
2:c237-0 liszt: 4169
2:c238-0 liszt: 4180
2:c239-0 liszt: 4191
2:c240-0 liszt: 3338
2:c241-0 liszt: 3348
2:c242-0 liszt: 4219
2:c243-0 liszt: 4230
2:c244-0 liszt: 4241
2:c245-0 liszt: 3379
2:c246-0 liszt: 2516
2:c247-0 liszt: 3392
2:c248-0 liszt: 3398
2:c249-0 liszt: 3404
2:c250-0 liszt: 2533
